Explainer implementation for LCPN #minor

ashishpatel16 opened the pull request:
This PR aims to add the implementation for LCPN explainer. The major changes in the code are as follows - 

- Since the code is structured in a way that `explain_with_xr` method relies on `get_traversed_nodes` for accessing the correct nodes for traversing the local classifiers, changes are only required in the `get_traversed_nodes` function to return traversed nodes depending on the Hiclass model provided.
- LCPN has the simplest way of returning the traversed nodes since each label in the prediction (`lcpn.predict()`) represents the node traversed at each hierarchy level. This is implemented in the corresponding method.
- This PR also adds new test case for LCPN, unfortunately, the tests are not generalizable, hence not parametrized. This is because the structure and shape of shap values is completely different for different Hiclass models, depending on the way they make predictions.

mirand863 commented:
@ashishpatel16 it would be great to have some documentation so the user can understand how the explainer works, how to use it and what is the expected output.

ashishpatel16 commented:
Hi @mirand863, this is the final version of the Explainer API for LCPN and LCPPN models. LCPL will follow soon. 

The changes made here are as follows -
- Updated readthedocs (Gallery of examples, Installation, API reference, Algorithm Overview)
- Updated the xarray Dataset by aligning the explanation object along required coordinate dimensions.
- Updated README to include an example of explaining Hiclass models.

I'm looking forward to your remarks. Hoping this will be the finally merged!

mirand863 reviewed (changes requested):
Hi @ashishpatel16 
The documentation is not building on read the docs. Please, add the required dependencies to the file [docs/requirements.txt](https://github.com/ashishpatel16/hiclass/blob/lcpn_explainer/docs/requirements.txt). I believe the tests are also not running due to dependencies not being installed in GitHub Actions. Please, also add the dependencies to "dev" in the file [setup.py](https://github.com/scikit-learn-contrib/hiclass/blob/main/setup.py) (you need to update your branch with the latest changes from the main)

ashishpatel16 commented on setup.py:
```diff
@@ -45,6 +45,8 @@
         "black==24.2.0",
         "pre-commit==2.20.0",
         "ray",
+        "shap==0.44.1",
```
@mirand863 I'm hardcoding this version of shap since there are some major changes in the API. The format and shape of `shap_values` has changed from version 0.44.1 -> [0.45.0](https://colab.research.google.com/drive/14sNFNveE1SPQvrEEAKKUSD1JPtFW6Y9S?usp=sharing). This was the reason why the previous PR was failing.

For quick comparison on the changed shapes of shap values, checkout [this](https://colab.research.google.com/drive/14sNFNveE1SPQvrEEAKKUSD1JPtFW6Y9S?usp=sharing) colab notebook.

ashishpatel16 commented:
Updated changes in this PR - 
- Merged the progress from main branch.
- [Colab](https://colab.research.google.com/drive/1wqSl1t_Qn2f62WNZQ48mdB0mNeu1XSF1?usp=sharing) notebook with updated code (refactored to use the hiclass.datasets module).
- Gallery of examples > LCPN [example](https://hiclass--108.org.readthedocs.build/en/108/auto_examples/plot_lcpn_explainer.html).
- Updated README with link to LCPN Colab notebook.

mirand863 commented:
Hi @ashishpatel16 

The colabs notebook does not work unless I update the git checkout to `git checkout --track origin/lcpn_explainer`. Can you please update it?

mirand863 commented on hiclass/Explainer.py:
```diff
@@ -189,6 +189,46 @@ def _get_traversed_nodes_lcppn(self, samples):
+        )
+
+        for level in range(1, traversals.shape[1]):
+            traversals[:, level] = np.core.defchararray.add(
```
Hi @ashishpatel16 
I was checking numpy's documentation and it seems defchararray is not the recommended way to do concatenations. Can you please update to numpy.char? Please see https://numpy.org/doc/stable/reference/generated/numpy.char.chararray.html and https://numpy.org/doc/stable/reference/routines.char.html#module-numpy.char

ashishpatel16 replied:
Hi @mirand863, Yes, you are correct! I have now used `numpy.char` for string concatenations as per the numpy documentation. 

Resolved by https://github.com/scikit-learn-contrib/hiclass/pull/108/commits/5aa9998664ec22843cae47c12c423087730930ab.

ashishpatel16 commented:
Hi @mirand863, 
Thanks for the review. I have updated the Colab notebook with the requested changes. Please check it and let me know if it works perfectly now! ✅

mirand863 commented:
> Hi @mirand863, Thanks for the review. I have updated the Colab notebook with the requested changes. Please check it and let me know if it works perfectly now! ✅

It works now, Thanks!
